staticcheck/util/directive_test.go

```diff
@@ -0,0 +1,84 @@
+package util
+
+import (
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"testing"
+
+	"honnef.co/go/tools/analysis/lint"
+)
+
+func TestAsIgnores(t *testing.T) {
+	filename := "sample.go"
+	sample := `
+package fail
+//lint:file-ignore ST3000 ignore all of these
+
+func alwaysTrue(a int) string {
+	//lint:ignore ST* ignore this
+	if a == a {
+		return "a"
+	}
+
+	return "b"
+}
+`
+	for _, test := range []struct {
+		desc, check     string
+		expectedIgnores []ignore
+	}{
+		{
+			desc:            "Should capture both directives",
+			check:           "ST3000",
+			expectedIgnores: []ignore{fileIgnore{File: filename}, lineIgnore{File: filename, Line: 7}},
+		},
+		{
+			desc:            "Should capture only ST*",
+			check:           "ST1000",
+			expectedIgnores: []ignore{lineIgnore{File: filename, Line: 7}},
+		},
+		{
+			desc:            "Should capture no directives",
+			check:           "QF1000",
+			expectedIgnores: nil,
+		},
+	} {
+		t.Run(test.desc, func(t *testing.T) {
+			fset := token.NewFileSet()
+			f, err := parser.ParseFile(fset, filename, sample, parser.ParseComments)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			lds := lint.ParseDirectives([]*ast.File{f}, fset)
+			if len(lds) != 2 {
+				t.Errorf("Expected 2 directives, got %d", len(lds))
+			}
+
+			igs := asIgnores(fset, test.check, lds)
+			if !slicesEquivalent(igs, test.expectedIgnores) {
+				t.Errorf("Expected ignores of %+v ignores, got %+v", test.expectedIgnores, igs)
+			}
+		})
+	}
+}
+
+func slicesEquivalent(actual, expected []ignore) bool {
+	if len(actual) != len(expected) {
+		return false
+	}
+
+	expectedAsMap := make(map[ignore]struct{})
+	for _, e := range expected {
+		expectedAsMap[e] = struct{}{}
+	}
+
+	for i := range actual {
+		if _, ok := expectedAsMap[actual[i]]; !ok {
+			return false
+		}
+	}
+
+	return true
```

Comment:
Nit: This approach would only validate that all elements in `actual` are in `expected`, not all elements in `expected` are in `actual`.

I do realize that this is only an assisting func for a test, so I don't mind merging this as-is if you don't wish to change it 🤗

Reply:
Updated to use a different approach; normally I'd use `github.com/stretchr/testify` for the assertions, but didn't want to add a new dependency to the project 😂